core/crypto/abac/attributes.go

```diff
@@ -0,0 +1,267 @@
+/*
+Copyright IBM Corp. 2016 All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+		 http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package abac
+
+import (
+	"bytes"
+	"crypto/x509"
+	"encoding/asn1"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+
+	pb "github.com/hyperledger/fabric/core/crypto/abac/proto"
+	"github.com/hyperledger/fabric/core/crypto/primitives"
+	"github.com/hyperledger/fabric/core/crypto/utils"
+
+	"github.com/golang/protobuf/proto"
+)
+
+var (
+	// TCertEncAttributesBase is the base ASN1 object identifier for attributes.
+	// When generating an extension to include the attribute an index will be
+	// appended to this Object Identifier.
+	TCertEncAttributesBase = asn1.ObjectIdentifier{1, 2, 3, 4, 5, 6}
+
+	// TCertAttributesHeaders is the ASN1 object identifier of attributes header.
+	TCertAttributesHeaders = asn1.ObjectIdentifier{1, 2, 3, 4, 5, 6, 9}
+
+	padding = []byte{255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255, 255}
+
+	//headerPrefix is the prefix used in the header exteion of the certificate.
+	headerPrefix = "00HEAD"
+
+	//HeaderAttributeName is the name used to derivate the K used to encrypt/decrypt the header.
+	HeaderAttributeName = "attributeHeader"
+)
+
+//ParseAttributesHeader parses a string and returns a map with the attributes.
+func ParseAttributesHeader(header string) (map[string]int, error) {
+	if !strings.HasPrefix(header, headerPrefix) {
+		return nil, errors.New("Invalid header")
+	}
+	headerBody := strings.Replace(header, headerPrefix, "", 1)
+	tokens := strings.Split(headerBody, "#")
+	result := make(map[string]int)
+
+	for _, token := range tokens {
+		pair := strings.Split(token, "->")
+
+		if len(pair) == 2 {
+			key := pair[0]
+			valueStr := pair[1]
+			value, err := strconv.Atoi(valueStr)
+			if err != nil {
+				return nil, err
+			}
+			result[key] = value
+		}
+	}
+
+	return result, nil
+}
+
+//ReadAttributeHeader read the header of the attributes.
+func ReadAttributeHeader(tcert *x509.Certificate, headerKey []byte) (map[string]int, bool, error) {
+	var err error
+	var headerRaw []byte
+	encrypted := false
+	if headerRaw, err = utils.GetCriticalExtension(tcert, TCertAttributesHeaders); err != nil {
```

Comment:
Yes, at this point the TCert is validated in the stub, that code is not part of this PR since we are splitting a bigger PR in smaller pieces.